### ISSUE TITLE
Could not write to env_meta.txt on Windows

### ISSUE THREAD

**Issue by Laudrin** [minetest/minetest#14619]
### Minetest version

```true
Minetest 5.8.0 (Windows)
Using Irrlicht 1.9.0mt13
Using LuaJIT 2.1.0-beta3
BUILD_TYPE=Release
RUN_IN_PLACE=1
USE_CURL=1
USE_GETTEXT=1
USE_SOUND=1
STATIC_SHAREDIR="."
STATIC_LOCALEDIR="locale"
```

### Operating system and version

Windows Version 10.0.19045 Build 19045

### CPU model

Intel(R) Core(TM) i3 CPU 530  @ 2.93GHz

### GPU model

NVIDIA GeForce GTX 1050 Ti

### Active renderer

OpenGL 4.6.0

### Summary

Minetest crashed when playing a single player game with a world I already played a lot of hours without any crashes.

The world has no other mods as the latest Mineclonia game version. I already played for months with the newest Minetest version and for weeks with the newest Mineclonia version (0.99.0) without this error or other crashes occurring. The error seems to be produced by the engine and not the game, this is why I am reporting this here.

On the SDD there is enough free space (170 GB), I did not change any permissions for the path. As virus protection, I use the internal MS Windows one, there was no report of any blocking.

```
2024-05-06 21:36:50: WARNING[Server]: Failed to write to file: C:\Users\Laudrin\Downloads\minetest-5.8.0-win64\bin\..\worlds\ANewBeginning\env_meta.txt
2024-05-06 21:36:50: ERROR[Server]: An unhandled exception occurred: "Couldn't save env meta"
2024-05-06 21:36:50: ERROR[Server]: In thread 10:
2024-05-06 21:36:50: ERROR[Server]: /home/stefan/mt-build/build/64/minetest/src/server.cpp:8c: virtual void* ServerThread::run(): A fatal error occurred: "Couldn't save env meta"
```

Maybe this is due to the hardcoded path `/home/stefan/mt-build/build/64/minetest/` which does not correspond to my Minetest path at all.

### Steps to reproduce

This was unexpected, as I did nothing unusual at that time.

**Comment by wsor4035** [comment]
seems like the usual corrupt env_meta.txt file, close/stop minetest, delete the file, and re start the world

**Comment by appgurueu** [comment]
(Note that sfan5's build path has nothing to do with this; it's just used for the debug info, probably because sfan made the build.)

**Comment by Laudrin** [comment]
Thank you, @wsor4035 & @appgurueu!

As I understand from looking up the use of the file, env_meta.txt contains global variables. Wouldn't that possibly break something in the game if I delete the file and restart the world?

This is the content of the file in the world I encountered the game crash in:

```
day_count = 475
lbm_introduction_times = :mcl_itemframes:convert_old_itemframes~0;:mcl_signs:update_old_signs~0;mcl_anvils:update_formspec_0_60_0~0;mcl_brewing:update_coolsneak~0;mcl_chests:replace_old_ender_form~0;mcl_chests:update_shulker_box_formspecs_0_72_0~0;mcl_core:replace_legacy_dry_grass_0_65_0~0;mcl_dispensers:update_formspecs_0_60_0~0;mcl_droppers:update_formspecs_0_60_0~0;mcl_furnaces:update_coolsneak~0;mcl_heads:convert_old_angled_heads~0;mcl_heads:convert_old_ceiling_heads~0;mcl_hoppers:update_formspec_0_60_0~0;mcl_loom:update_coolsneak~0;mcl_mushrooms:replace_legacy_mushroom_caps~0;mcl_ocean:fix_incorrect_seagrass~0;mcl_ocean:kelp_init_83~0;mcl_smithing_table:update_coolsneak~0;mcl_trees:leaves_param2_update~0;
lbm_introduction_times_version = 1
last_clear_objects_time = 0
time_of_day = 10424
game_time = 404073
EnvArgsEnd
```

Today I got two more crashes with the same error message.

Is this maybe related to the second line in the file? Maybe the syntax is off? Should `lbm_introduction_times = ` be followed by a value and then a semicolon between this variable and the next instead of a colon `:mcl_itemframes:convert_old_itemframes~0`. Is this something to report to the developers of the Mineclonia game instead of here or is it unrelated to MCLA?

**Comment by appgurueu** [comment]
> As I understand from looking up the use of the file, env_meta.txt contains global variables. Wouldn't that possibly break something in the game if I delete the file and restart the world?

Yes. I'd recommend backing up the world. As you can see in your example though, what it stores is quite limited (time, some LBM related info); often it's possible to delete it without issues. This depends on the game though. I don't know if this would cause any issues in Mineclonia.

I don't think the syntax is wrong. The ABMs probably are registered using names starting with a leading colon. (Probably someone assumed that the same conventions / requirements as for nodenames were enforced here, which is not quite the case.)

So indeed deleting it is a "last resort" and a dirty hack; "repairing" it would be a preferable solution, if possible.

> Is this something to report to the developers of the Mineclonia game instead of here or is it unrelated to MCLA?

Most probably unrelated to Mineclonia.

---

I don't know what causes this, but as the error message suggests, it is most probably an issue with saving this file. I don't know what that would be caused by, but there are a plethora of possible reasons, many out of Minetest's control. Have you ensured that you're not running out of disk space, that the drive the world resides on is mounted, that Minetest has sufficient permissions, that there's no AV or similar interfering (note: I don't think AV programs will always tell you when they interfere), etc.? You could try using something like [x64dbg](https://x64dbg.com/) to debug this (disclaimer: I'm not very familiar with debugging on Windows).

**Comment by sfan5** [comment]
It would be useful if you could go [here](https://github.com/minetest/minetest/actions/runs/9100456587#artifacts), download the "mingw64" artifact and try to reproduce the same errors.
Then if it happens again, paste the text from debug.txt again.

**Comment by Laudrin** [comment]
> It would be useful if you could go [here](https://github.com/minetest/minetest/actions/runs/9100456587#artifacts), download the "mingw64" artifact and try to reproduce the same errors. Then if it happens again, paste the text from debug.txt again.

Sorry for the late reply, @sfan5, I did not find the time before, but today I downloaded the artifact you suggested and let it run for a while.

It seems more stable then my former MT version, as that one tends to crash if I start applications like a browser in background. It seems to have something to do with other programs using a lot of memory and resources, which should slow down MT, but not crash it.

After a while of letting this mingw64 version run and starting and closing lot of apps in the background (which would surely have crashed my "normal" MT version), this one finally crashed, too. 😥

`2024-06-28 13:34:52: WARNING[Server]: active block modifiers took 217ms (processed 94 of 485 active blocks)
2024-06-28 13:34:52: ERROR[Server]: Failed to overwrite "C:\Users\Laudrin\Downloads\mingw64\minetest-5.9.0-dev-win64\bin\..\worlds\VerdammtNeu!\env_meta.txt": Zugriff verweigert
2024-06-28 13:34:52: ERROR[Server]: An unhandled exception occurred: "Couldn't save env meta"
2024-06-28 13:34:52: ERROR[Server]: In thread 9084:
2024-06-28 13:34:52: ERROR[Server]: /home/runner/work/minetest/minetest/src/server.cpp:96: virtual void *ServerThread::run(): A fatal error occurred: "Couldn't save env meta"
`

Debugging it with Visual Studio I got the error message:

`Ausnahmefehler bei 0x00007FFA69FC286E (ucrtbase.dll) in minetest.exe: Schwerwiegende Beendigung des Programms angefordert`
 which translated is:
`Exception error at 0x00007FFA69FC286E (ucrtbase.dll) in minetest.exe: Fatal termination of the programme requested`

As I don't know how to use the debugger (MS Visual Studio debugger or the one suggested by @appgurueu), I cannot get more useful information from here on.

But if you want me to do some more tests, please tell me what I could do and I will gladly try to get as much information as I could that this issue could be fixed! 😄

**Comment by appgurueu** [comment]
> ERROR[Server]: Failed to overwrite "C:\Users\Laudrin\Downloads\mingw64\minetest-5.9.0-dev-win64\bin\..\worlds\VerdammtNeu!\env_meta.txt": Zugriff verweigert

What's up with this "access denied"? Are your file permissions correct? Is there perhaps some AV program interfering? Are you perhaps running out of disk space, causing I/O to fail (though that should probably have a better error message than "access denied")?

> It seems to have something to do with other programs using a lot of memory and resources, which should slow down MT, but not crash it.

Is it possible that you were running out of memory? It is very plausible that the code is written in such a way that it doesn't deal with OOM appropriately.

**Comment by Laudrin** [comment]
> > ERROR[Server]: Failed to overwrite "C:\Users\Laudrin\Downloads\mingw64\minetest-5.9.0-dev-win64\bin..\worlds\VerdammtNeu!\env_meta.txt": Zugriff verweigert
> 
> What's up with this "access denied"? Are your file permissions correct? Is there perhaps some AV program interfering? Are you perhaps running out of disk space, causing I/O to fail (though that should probably have a better error message than "access denied")?

No, the files are on a 2 TB SSD with ca. 90 GB free space, no other program reports errors or do such frequent crashes like MT does.

File permissions are fine, too. AV is build in MS AV, no external antivirus.

> Is it possible that you were running out of memory? It is very plausible that the code is written in such a way that it doesn't deal with OOM appropriately.

I will try to monitor my free memory. I have 16 GB, but of course other programs could fill the memory. Theres a swapfile, too, that's why I said it then should slow down but not crash imho.

I will take a look if I find a tool to do so. If you know a free memory monitoring tool showing free physical memory in the taskbar or as an overlay on the screen, I would be glad if you could point me to one.

Or if I could use some debug level option or something else in MT to see if the available memory gets low?

**Comment by Laudrin** [comment]
I checked the Windows event log:

```
Application Error, Event-ID 1000

Name of the faulty application: minetest.exe, version: 5.9.0.0, timestamp: 0x6644f93d
Name of the faulty module: ucrtbase.dll, version: 10.0.19041.3636, timestamp: 0x81cf5d89
Exception code: 0xc0000409
Fehleroffset: 0x000000000007286e
ID of the faulty process: 0x6b3c
Start time of the faulty application: 0x01dace21c4c58c38
Path of the faulty application: C:\Users\Laudrin\Downloads\mingw64\minetest-5.9.0-dev-win64\bin\minetest.exe
Path of the faulty module: C:\WINDOWS\System32\ucrtbase.dll
Berichtskennung: 2c3efd9e-5755-4e16-8146-9342ba44a1c1
Full name of the faulty package: 
Application ID that is relative to the faulty package:
```

Sorry,  this is a (re-)translation from German.

Is ucrtbase.dll the module causing the problem, or just the "messenger" not to be shot?

And another thing: I deleted env_meta.txt multiple times, but I cannot see a difference in the frequency of the applications crashes.

**Comment by Laudrin** [comment]
![Minetest Crash](https://github.com/minetest/minetest/assets/36445989/be934e59-6d4d-45a4-abfe-e45bd3a25b77)

Not sure if I caught the right moment and if Windows task manager gives enough information.

But it seems memory is not the problem as it was at 9.7 GB (of 16 GB) the whole time, according to task manager. Now after the program crashed and was closed, it's at about 8.9 GB.

CPU was at full load, could that be the problem?

**Comment by appgurueu** [comment]
> CPU was at full load, could that be the problem?

That could create problems related to scheduling, but they shouldn't manifest like this.

---

I googled your exception code [and it's apparently a stack buffer overrun](https://stackoverflow.com/questions/23409809/how-do-you-diagnose-the-exception-code-0xc0000409-on-windows).

I see no sensible other way to proceed than either:

* Running Minetest with a debugger attached to it to get a stack trace. You will need a debug build for this.
* Compiling Minetest with an address sanitizer so that the stack buffer overrun trips that and we get a stack trace.

We have a [basic guide](https://github.com/minetest/minetest/blob/master/doc/compiling/windows.md) for compiling on Windows. There is also [this video](https://www.youtube.com/watch?v=B4QnlJozFoM) by jordan4ibanez.

**Comment by sfan5** [comment]
> I googled your exception code [and it's apparently a stack buffer overrun](https://stackoverflow.com/questions/23409809/how-do-you-diagnose-the-exception-code-0xc0000409-on-windows).

following a few further links it actually just means `abort` was called: https://devblogs.microsoft.com/oldnewthing/20190108-00/?p=100655
I think we're at a dead end here unless someone does a deep dive with some advances tools like [Procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) to locate where exactly the error is produced.

**Comment by Laudrin** [comment]
@appgurueu and @sfan5: Thank you for trying to find the cause for this issue!

For a few days now, there were no further crashes. 🎉 But as I don't understand why it could happen again any day. 😕 The only thing I changed was updating some Windows programs, letting Windows install the usual updates and updating some mods when updates were available via the internal MT updater.

I did not change which programs I used in the background / switching between them or the way I used those programs. So it is still a mystery.

I will observe this further and as soon as I have new information, post here.

For now, thank you all (including of course @wsor4035) for trying to help me with this issue! 🙂
